### PR TITLE
Potato weiping

### PR DESCRIPTION
**Pull request by weipinglu**

### CODE REVIEW

File: src/Potato/Data/Text/Zipper2.hs [hunk 1]
```diff
@@ -155,9 +155,14 @@ up    (TextZipper (x:xs) b [y] a la) = TextZipper (NE.tail lb) b' [] a' ((by <>
 
 -- | Move the cursor down one logical line (clearing the selection)
 down :: TextZipper -> TextZipper
-down (TextZipper lb b [] "" []) = TextZipper lb b [] "" []
-down (TextZipper lb b [] "" (x:xs)) = TextZipper (lb <> [b]) x [] "" (NE.tail a) where
-    a = x NE.:| xs
+down (TextZipper lb b [] a [])    = TextZipper lb (b <> a) [] "" []
+down (TextZipper lb b [] a (x:xs) = TextZipper ((b <> a):lb) b' [] a' xs where
+        (b', a') = T.splitAt (T.length x) x
```
Comment (minimapletinytools) [inline src/Potato/Data/Text/Zipper2.hs]: I think you meant `length b`

File: src/Potato/Data/Text/Zipper2.hs [hunk 2]
```diff
@@ -155,9 +155,14 @@ up    (TextZipper (x:xs) b [y] a la) = TextZipper (NE.tail lb) b' [] a' ((by <>
...
+down (TextZipper lb b [y] a [] = TextZipper lb (b <> y <> a) [] "" []
+down (TextZipper lb b [y] a (x:xs) = TextZipper ((by:lb) b' [] a' xs where
+        (b', a') = T.splitAt (T.length by) x
```
Comment (minimapletinytools) [inline src/Potato/Data/Text/Zipper2.hs]: where is `by` defined?

File: src/Potato/Data/Text/Zipper2.hs [hunk 3]
```diff
@@ -155,9 +155,14 @@ up    (TextZipper (x:xs) b [y] a la) = TextZipper (NE.tail lb) b' [] a' ((by <>
...
+down (TextZipper lb b (y:(ys:yss)) a la = TextZipper ((by:lb) b' [] a' xs where
+        (b', a') = T.splitAt (T.length by) x
```
Comment (pdlla) [inline src/Potato/Data/Text/Zipper2.hs:165]: where is x defined?

**Review by minimapletinytools — COMMENTED** [review]
I don't think your current implementation will compile. Some variables look undefined.

File: src/Potato/Data/Text/Zipper2.hs [hunk 4]
```diff
@@ -142,11 +141,31 @@ rightLeftWord = undefined
 
 -- | Move the cursor up one logical line (clearing the selection)
 up :: TextZipper -> TextZipper
-up z@(TextZipper lb b s a la) = undefined
+up    (TextZipper []     b [] a la) = TextZipper [] "" [] (b <> a) la
+up    (TextZipper (x:xs) b [] a la) = TextZipper (NE.tail lb) b' [] a' ((b <> a):la) where
+        lb = x NE.:| xs
+        (b', a') = T.splitAt (T.length b) x
+up    (TextZipper [] b [y] a la) = TextZipper [] "" [] (b <> y <> a) la
+up    (TextZipper lb b (y:(ys:yss)) a la) = TextZipper lb "" [] (b <> y) (NE.init ys' <> [NE.last ys' <> a] <> la) where
```
Comment (minimapletinytools) [inline src/Potato/Data/Text/Zipper2.hs]: isn't `NE.init ys' <> [NE.last ys' <> a]` just `ys <> yss <> a`

Comment (minimapletinytools) [reply src/Potato/Data/Text/Zipper2.hs]: I don't think I agree with this interpretation of `up` in this case, this is looks to be the same as `home`

File: src/Potato/Data/Text/Zipper2.hs [hunk 5]
```diff
@@ -142,11 +141,31 @@ rightLeftWord = undefined
 
 -- | Move the cursor up one logical line (clearing the selection)
 up :: TextZipper -> TextZipper
-up z@(TextZipper lb b s a la) = undefined
```
Comment (minimapletinytools) [inline src/Potato/Data/Text/Zipper2.hs:145]: I think you can make several of these cases a lot easier if you deselect first and then recursively call up. Then all of the `(y:(ys:yss))` and `[y]` cases are handled by the `[]` cases.

File: src/Potato/Data/Text/Zipper2.hs [hunk 6]
```diff
@@ -142,11 +141,31 @@ rightLeftWord = undefined
...
+        ys' = ys NE.:| yss
+up    (TextZipper (x:xs) b [y] a la) = TextZipper (NE.tail lb) b' [] a' ((by <> a):la) where
+        lb = x NE.:| xs
+        by = b <> y
+        (b', a') = T.splitAt (T.length by) x
 
 -- | Move the cursor down one logical line (clearing the selection)
 down :: TextZipper -> TextZipper
-down z@(TextZipper lb b s a la) = undefined
+down (TextZipper lb b [] a [])     = TextZipper lb (b <> a) [] "" []
```
Comment (minimapletinytools) [inline src/Potato/Data/Text/Zipper2.hs]: Do you think you can also simply this implementation too by deselcting first and then calling `down` recursively?

File: src/Potato/Data/Text/Zipper2.hs [hunk 7]
```diff
@@ -172,7 +191,11 @@ end (TextZipper lb b (x:(xs:xss)) a la) =TextZipper (lb <> ([b <> x] <> NE.init
 
 -- | Move the cursor to the top of the document (clearing the selection)
 top :: TextZipper -> TextZipper
-top (TextZipper lb b s a la) = undefined
+top tz@(TextZipper [] "" [] a la) = tz
+top (TextZipper [x]    "" [] a la) = TextZipper [] "" [] x (a:la)
+top (TextZipper (x:xs) "" [] a la) = TextZipper [] "" [] (NE.last xs') ((reverse $ NE.init xs') <> (a:la)) where
+    xs' = x NE.:| xs
+top tz = top $ home tz
```
Comment (minimapletinytools) [inline src/Potato/Data/Text/Zipper2.hs:191]: nice !

File: src/Potato/Data/Text/Zipper2.hs [hunk 8]
```diff
@@ -172,7 +191,11 @@ end (TextZipper lb b (x:(xs:xss)) a la) =TextZipper (lb <> ([b <> x] <> NE.init
 
 -- | Move the cursor to the top of the document (clearing the selection)
 top :: TextZipper -> TextZipper
-top (TextZipper lb b s a la) = undefined
+top tz@(TextZipper [] "" [] a la) = tz
+top (TextZipper [x]    "" [] a la) = TextZipper [] "" [] x (a:la)
+top (TextZipper (x:xs) "" [] a la) = TextZipper [] "" [] (NE.last xs') ((reverse $ NE.init xs') <> (a:la)) where
```
Comment (minimapletinytools) [inline src/Potato/Data/Text/Zipper2.hs:189]: nice!

File: test/Potato/Data/Text/Zipper2Spec.hs [hunk 9]
```diff
@@ -109,3 +104,85 @@ spec =
                     ["blah blah blah"]
     home tz_0 `shouldBe` tz_1h
     end tz_0 `shouldBe` tz_1e
+
+  it "top examples" $ do
+    top (TextZipper [] "" [] "dd d" ["a","b"]) `shouldBe`
```
Comment (minimapletinytools) [inline test/Potato/Data/Text/Zipper2Spec.hs:109]: nice alignment!

File: test/Potato/Data/Text/Zipper2Spec.hs [hunk 10]
```diff
@@ -109,3 +104,85 @@ spec =
...
+    up (TextZipper ["cat dog"] "abfeeeere " [] "dd d" ["a","b"]) `shouldBe` 
+        (TextZipper [] "cat dog" [] "" ["abfeeeere dd d", "a","b"])
+    up (TextZipper (reverse ["cat dog", "bird_sky"]) "ab " [] "dd d" ["a","b"]) `shouldBe` 
+        (TextZipper ["cat dog"] "bir" [] "d_sky" ["ab dd d", "a","b"])
+    up (TextZipper [] "ab " ["story"] "dd d" ["a","b"]) `shouldBe` 
+        (TextZipper [] "" [] "ab storydd d" ["a","b"])
+    up (TextZipper [] "ab " ["story", "cool"] "dd d" ["a","b"]) `shouldBe` 
+        (TextZipper [] "" [] "ab story" ["cooldd d", "a","b"])
+    up (TextZipper ["a", "b"] "ab " ["story", "cool"] "dd d" ["a","b"]) `shouldBe` 
+        (TextZipper ["a", "b"] "" [] "ab story" ["cooldd d", "a","b"])
```
Comment (minimapletinytools) [inline test/Potato/Data/Text/Zipper2Spec.hs]: shouldn't this be 

`TextZipper ["a"] "b" [] "ab story" ["cooldd d", "a","b"]`

File: src/Potato/Data/Text/Zipper2.hs [hunk 11]
```diff
@@ -140,13 +139,30 @@ rightWord = undefined
...
+deselect    (TextZipper lb b [x]          a la) = TextZipper lb (b <> x) [] a la
+deselect    (TextZipper lb b (x:(xs:xss)) a la) = TextZipper ((reverse $ NE.init xs') <> [b <> x] <> lb) (NE.last xs') [] a la where
+                xs' = xs NE.:| xss
+
 -- | Move the cursor up one logical line (clearing the selection)
-up :: TextZipper -> TextZipper
-up z@(TextZipper lb b s a la) = undefined
+up ::  TextZipper -> TextZipper
+up    (TextZipper [] b [] a la) = TextZipper [] "" [] (b <> a) la
+up    (TextZipper [x] b [] a la) = TextZipper [] b' [] a' ([b <> a] <> la) where
```
Comment (minimapletinytools) [inline src/Potato/Data/Text/Zipper2.hs]: this case is covered by `up    (TextZipper (x:xs) b [] a la)` case

File: src/Potato/Data/Text/Zipper2.hs [hunk 12]
```diff
@@ -140,13 +139,30 @@ rightWord = undefined
...
+        (b', a') = T.splitAt (T.length b) x
+up    (TextZipper (x:xs) b [] a la) = TextZipper xs b' [] a' ([b <> a] <> la) where
+        (b', a') = T.splitAt (T.length b) x
+up tz@(TextZipper lb b s a la) = up $ deselect tz
 
 -- | Move the cursor down one logical line (clearing the selection)
 down :: TextZipper -> TextZipper
-down z@(TextZipper lb b s a la) = undefined
+down    (TextZipper lb b []  a []) = TextZipper lb (b <> a) [] "" []
+down    (TextZipper lb b [] a [x]) = TextZipper ([b <> a] <> lb) b' [] a' [] where
```
Comment (minimapletinytools) [inline src/Potato/Data/Text/Zipper2.hs]: this case is covered by `down    (TextZipper lb b [] a (x:xs))` case

File: src/Potato/Data/Text/Zipper2.hs [hunk 13]
```diff
@@ -172,7 +188,11 @@ end (TextZipper lb b (x:(xs:xss)) a la) =TextZipper (lb <> ([b <> x] <> NE.init
 
 -- | Move the cursor to the top of the document (clearing the selection)
 top :: TextZipper -> TextZipper
-top (TextZipper lb b s a la) = undefined
+top tz@(TextZipper [] "" [] a la) = tz
```
Comment (minimapletinytools) [inline src/Potato/Data/Text/Zipper2.hs]: This is fine, you can also do 
`top tz@(TextZipper [] "" [] _ _)` in this case

**Review by minimapletinytools — COMMENTED** [review]
LGTM, looks like you can simplify a few of the functions even more!

File: src/Potato/Data/Text/Zipper2.hs (hunk 12 above)
Comment (weipinglu) [reply src/Potato/Data/Text/Zipper2.hs]: All review comments addressed

**Comment by weipinglu** [comment]
All comments addressed in the latest commit

**Review by minimapletinytools — APPROVED** [review]
looking very good!

**Pull request merged**